packages/ai-llamafile/src/browser/llamafile-frontend-application-contribution.ts

```diff
@@ -0,0 +1,63 @@
+// *****************************************************************************
+// Copyright (C) 2024 EclipseSource GmbH.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0.
+//
+// This Source Code may also be made available under the following Secondary
+// Licenses when the conditions for such availability set forth in the Eclipse
+// Public License v. 2.0 are satisfied: GNU General Public License, version 2
+// with the GNU Classpath Exception which is available at
+// https://www.gnu.org/software/classpath/license.html.
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import { LanguageModelRegistry } from '@theia/ai-core';
+import { FrontendApplicationContribution, PreferenceService } from '@theia/core/lib/browser';
+import { inject, injectable } from '@theia/core/shared/inversify';
+import { LlamafileLanguageModel } from '../common/llamafile-language-model';
+import { LlamafileServerManager } from '../common/llamafile-server-manager';
+import { LlamafileEntry, PREFERENCE_LLAMAFILE } from './llamafile-preferences';
+
+@injectable()
+export class LlamafileFrontendApplicationContribution implements FrontendApplicationContribution {
+
+    @inject(PreferenceService)
+    protected preferenceService: PreferenceService;
+
+    @inject(LanguageModelRegistry)
+    protected languageModelRegistry: LanguageModelRegistry;
+
+    @inject(LlamafileServerManager)
+    protected llamafileServerManager: LlamafileServerManager;
+
+    onStart(): void {
+        this.preferenceService.ready.then(() => {
+            const llamafiles = this.preferenceService.get<LlamafileEntry[]>(PREFERENCE_LLAMAFILE, []);
+
+            const models = llamafiles.map(llamafile =>
+                LlamafileLanguageModel.createNewLlamafileLanguageModel(llamafile.name, llamafile.uri, llamafile.port, this.llamafileServerManager));
+            this.languageModelRegistry.addLanguageModels(models);
+
+            this.preferenceService.onPreferenceChanged(event => {
+                if (event.preferenceName === PREFERENCE_LLAMAFILE) {
+                    // old models in the preference
+                    const oldModels = new Set((event.oldValue as LlamafileEntry[]).map(v => v.name));
```

Comment:
`oldValue` will be `undefined` when the preference was not set before. In that case the `new Set` will actually result in `undefined` too and `.map` will throw an error.